---
title: "Ex18 Phospho proteomics"
output: html_document
---

Comparing across Early, Late. Comparing to G treatment by state.


```{r include=FALSE}
library(MSnSet.utils)
library(clusterProfiler)
library(org.Hs.eg.db)
library(dplyr)
library(ggplot2)
library(readxl)
library(pheatmap)
library(msigdbr)
library(stringr)
library(VennDiagram)

global_crosstab <- read.table("../../proteomics/data/global_data/ptrc_ex18_crosstab_global_gene_corrected.txt")
phospho_crosstab <- read.table("../../proteomics/data/phospho_data/ptrc_ex18_crosstab_phospho_siteID_corrected.txt")
phospho_prot_crosstab <- read.table("../../proteomics/data/phospho_data/ptrc_ex18_crosstab_phospho_SiteID_prot_corrected.txt")
meta <- read_excel("../../proteomics/data/Exp18_metadata_01-14-2021.xlsx") %>%
  filter(!is.na(Index)) %>%
  mutate(Index = as.character(Index),
         Plex = as.character(Plex),
         comp1_ = paste(State, Treatment, sep = "_and_")) %>% as.data.frame()
rownames(meta) <- meta$Index

colnames(global_crosstab) <- sub("^X", "", colnames(global_crosstab))
colnames(phospho_crosstab) <- sub("^X", "", colnames(phospho_crosstab))
colnames(phospho_prot_crosstab) <- sub("^data.", "", colnames(phospho_prot_crosstab))


m_global <- MSnSet(exprs = global_crosstab %>% as.matrix(), 
                   pData = meta[colnames(global_crosstab), ])

m_phospho <- MSnSet(exprs = phospho_crosstab %>% as.matrix(), 
                    pData = meta[colnames(phospho_crosstab), ])

# ## data only
# phospho_prot_mat <- phospho_prot_crosstab[, intersect(as.character(1:47), colnames(phospho_prot_crosstab))]
# mapping_df <- phospho_prot_crosstab %>%
#   mutate(feature = rownames(.)) %>%
#   dplyr::select(feature, Peptide_kstar, Protein_kstar, SiteID)
# 
# m_phospho_prot <- MSnSet(exprs = phospho_prot_mat %>% as.matrix(), 
#                          pData = meta[colnames(phospho_crosstab), ])

```


```{r}
# diffexp_results_list <- c(list.files("Tables", "^diffexp_.*_Phospho.txt", full.names = T),
#                           list.files("Tables", "^diffexp_.*_Global.txt", full.names = T))
diffexp_results_list <- c(list.files("../treatments_vs_parental/Tables", "^diffexp_.*_Global.txt", full.names = T))

diffexp_results_global <- data.frame()
for (diff_result in diffexp_results_list){
  xx <- read.table(diff_result, sep = "\t") %>% dplyr::select(feature, logFC, welch_adj, contrast)
  diffexp_results_global <- rbind(diffexp_results_global, xx)
}

diffexp_results_list <- c(list.files("../treatments_vs_parental/Tables", "^diffexp_.*_Phospho.txt", full.names = T))

diffexp_results_phospho <- data.frame()
for (diff_result in diffexp_results_list){
  xx <- read.table(diff_result, sep = "\t") %>% dplyr::select(feature, logFC, welch_adj, contrast)
  diffexp_results_phospho <- rbind(diffexp_results_phospho, xx)
}


```


```{r}
## Counts global

count_df <- diffexp_results_global %>%
  filter(welch_adj < 0.05) %>%
  mutate(Treatment = sub("^.*_and_(.*)-.*$", "\\1", contrast),
         State = sub("(^.*)_and_.*-.*$", "\\1", contrast)) %>%
  group_by(feature, Treatment) %>%
  mutate(total_states = n(),
         agree = case_when(total_states > 1 ~ sign(prod(logFC)),
                           TRUE ~ 1)) %>%
  mutate(agree = case_when(agree == 1 ~ "Agree across state",
                           TRUE ~ "Disagree across state")) %>%
  ungroup() %>% group_by(Treatment, State) %>%
  mutate(total_combo = n()) %>% ungroup()

plot_df <- count_df %>%
  select(Treatment, State, counts = total_combo) %>% 
  unique()

xx <- count_df %>%
  filter(total_states == 2,
         State == "Early") %>%
  select(feature, Treatment, agree) %>%
  unique() %>%
  group_by(Treatment, agree) %>%
  summarize(counts = n()) %>%
  select(Treatment, State = agree, counts) 

plot_df <- rbind(plot_df, xx) %>%
  rbind(data.frame(Treatment = c("GD", "GV", "GVD"), 
                   State = "Disagree across state", counts = 0)) %>%
  mutate(ID = paste(Treatment, State),
         State = factor(State, levels = c("Early", "Late", "Agree across state", "Disagree across state"))) %>%
  select(Counts = counts, everything())

p <- ggplot(plot_df, aes(x = Treatment, fill = State, y = Counts)) + 
  geom_bar(stat = 'identity', width = 0.65, position = "dodge") +
  scale_fill_manual(values = c("cadetblue3", "dodgerblue2", "olivedrab3", "orangered1")) +
  ylab("Differentially expressed genes") +
  theme(text = element_text(size = 16))
  
ggsave(plot = p, filename = "global_diffexp_barplot_summary.png", height = 7, width = 9)

## Counts Phospho

count_df <- diffexp_results_phospho %>%
  filter(welch_adj < 0.05) %>%
  mutate(Treatment = sub("^.*_and_(.*)-.*$", "\\1", contrast),
         State = sub("(^.*)_and_.*-.*$", "\\1", contrast)) %>%
  group_by(feature, Treatment) %>%
  mutate(total_states = n(),
         agree = case_when(total_states > 1 ~ sign(prod(logFC)),
                           TRUE ~ 1)) %>%
  mutate(agree = case_when(agree == 1 ~ "Agree across state",
                           TRUE ~ "Disagree across state")) %>%
  ungroup() %>% group_by(Treatment, State) %>%
  mutate(total_combo = n()) %>% ungroup()

plot_df <- count_df %>%
  select(Treatment, State, counts = total_combo) %>% 
  unique()

xx <- count_df %>%
  filter(total_states == 2,
         State == "Early") %>%
  select(feature, Treatment, agree) %>%
  unique() %>%
  group_by(Treatment, agree) %>%
  summarize(counts = n()) %>%
  select(Treatment, State = agree, counts) 

plot_df <- rbind(plot_df, xx) %>%
  rbind(data.frame(Treatment = c("GD", "GV", "GVD"), 
                   State = "Disagree across state", counts = 0)) %>%
  mutate(ID = paste(Treatment, State),
         State = factor(State, levels = c("Early", "Late", "Agree across state", "Disagree across state"))) %>%
  select(Counts = counts, everything())

p <- ggplot(plot_df, aes(x = Treatment, fill = State, y = Counts)) + 
  geom_bar(stat = 'identity', width = 0.65, position = "dodge") +
  scale_fill_manual(values = c("cadetblue3", "dodgerblue2", "olivedrab3", "orangered1")) +
  ylab("Differentially expressed genes") +
  theme(text = element_text(size = 16))
  
ggsave(plot = p, filename = "phospho_diffexp_barplot_summary.png", height = 7, width = 9)
  

```


```{r}
compare_groups_phospho <- function(group1, group2, results_df, pval_cutoff = 0.34){
  name1 <- gsub("_and_", " ", group1) %>% gsub("-", " vs ", .) %>% sub(" none", "", .)
  name2 <- gsub("_and_", " ", group2) %>% gsub("-", " vs ", .) %>% sub(" none", "", .)
  name1 <- sub(" vs Parental", "", name1)
  name2 <- sub(" vs Parental", "", name2)
  N_features <- results_df %>% filter(contrast == group1) %>% nrow()
  all_features <- unique(results_df$feature)
  group1_df <- results_df %>%
    filter(contrast == group1,
           welch_adj < 0.05) %>%
    dplyr::select(feature, logFC)
  
  group2_df <- results_df %>%
    filter(contrast == group2,
           welch_adj < 0.05) %>%
    dplyr::select(feature, logFC)
  
  common_features <- intersect(group1_df$feature, group2_df$feature)
  pval <- hypergeometric_pval <- phyper(q = length(common_features), m = nrow(group1_df), 
                                        n = N_features - nrow(group1_df), k = nrow(group2_df), lower.tail = F)
  combined <- rbind(group1_df %>% filter(feature %in% common_features), 
                    group2_df %>% filter(feature %in% common_features)) %>%
    as.data.frame() %>%
    group_by(feature) %>%
    mutate(logFC = sign(logFC), 
           agree = sign(prod(logFC)) == 1)
  
  disagree <- combined %>% filter(!agree) %>%
    pull(feature) %>% unique()
  
  disagree_df <- rbind(group1_df %>% filter(feature %in% disagree) %>% mutate(group = group1),
                      group2_df %>% filter(feature %in% disagree) %>% mutate(group = group2)) %>%
    mutate(group = sub("-Parental_and_none$", "", group),
           group = gsub("_", " ", group),
           color = case_when(logFC > 0 ~ "red",
                             logFC < 0 ~ "blue")) 
  
  # disagree_plot <- ggplot(disagree_df, aes(x = logFC, y = feature, fill = color)) + 
  #   geom_bar(stat = 'identity') + 
  #   facet_grid(. ~ group) + 
  #   guides(fill = "none") #+ theme(text = element_text(size = 21))
  
  # ggsave(plot = disagree_plot, filename = paste0("disagree_plot_phospho", group1, "_", group2, ".png"))
  
  combined <- combined %>%
    filter(agree) %>%
    dplyr::slice(1)


  venn.diagram(
    x = list(group1_df$feature, group2_df$feature),
    category.names = c(name1, name2),
    filename = paste("overlap_venn", gsub(" ", "_", tolower(name1)), "and",
                     gsub(" ", "_", tolower(name2)), "vs_parental_phospho.png", sep = "_"),
    output = FALSE,

    # Output features
    imagetype="png" ,
    height = 700 ,
    width = 700,
    resolution = 300,
    lwd = 1,
    col=c("#440154ff", '#21908dff'),
    fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3)),
    cex = 0.65,
    fontfamily = "sans",
    cat.cex = 0.7,
    cat.default.pos = "outer",
    cat.pos = c(-165, 15),
    cat.dist = c(0.05, 0.05),
    cat.col = c("#440154ff", '#21908dff'),
    main = paste0("pvalue = ", pval),
    main.cex = 0.69
  )
  
  plot_title <- paste("Sites active in", sub(" vs Parental", "", name1), "and", sub(" vs Parental", "", name2), "vs Parental")
  filename = paste0("over_representation_ksdb_", gsub(" ", "_", tolower(plot_title)), ".png")
  
  helper_s <- function(phospho_name){
    out = list()
    while(grepl("[a-z]", phospho_name)){
      phospho_name <- sub("[a-z]{1}$", "", phospho_name)
      ending <- sub("^.*[a-z]{1}([A-Z, 0-9]+)$", "\\1", phospho_name)
      out <- append(out, ending)
      phospho_name <- sub(ending, "", phospho_name)
    }
    protein <- sub("-[A-Z]{1}[0-9]+$", "", out[[length(out)]])
    if (length(out) > 1){
      out <- c(paste(protein, out[1:(length(out))-1], sep = "-"), out[[length(out)]])
    } else{
      out <- out[[1]]
    }
    return(out)
  }
  
  helper <- function(features){
    common_sites <- c()
    for (feature in features){
      if (!grepl("NULL$", feature)){
        common_sites <- c(common_sites, helper_s(feature))
      }
    }
    return(common_sites)
  }
  
  clustList <- list('Over-expressed' = combined %>% filter(logFC == 1) %>% filter(agree) %>% pull(feature) %>% helper(),
                    'Under-expressed' = combined %>% filter(logFC == -1) %>% pull(feature) %>% helper(),
                    'All features' = all_features %>% helper())
  

  # clustList <- lapply(clustList, intersect, mappedRkeys(org.Hs.egSYMBOL2EG))
  # clustEG <- lapply(clustList, function(x) sapply(as.list(org.Hs.egSYMBOL2EG[x]), `[`, 1))
  
  KSDB <- read.csv(system.file('PSP&NetworKIN_Kinase_Substrate_Dataset_July2016.csv',
                             package='amlresistancenetworks'),stringsAsFactors = FALSE) %>%
    filter(Source == "PhosphoSitePlus") %>%
    dplyr::mutate(site_name = paste(SUB_GENE, SUB_MOD_RSD, sep = "-")) %>%
    dplyr::select(gs_name = GENE, entrez_gene = site_name)

  xx <- compareCluster(geneCluster = clustList[1:2], fun = enricher,
                       TERM2GENE = KSDB, TERM2NAME = NA, 
                       pvalueCutoff = 1, qvalueCutoff = 1,
                       minGSSize = 1, maxGSSize = 600,
                       pAdjustMethod = "BH",
                       universe=clustList[[3]])
  # # plot(xx, colorBy="qvalue")
  p <- try(dotplot(xx, showCategory = 7) +
    scale_color_gradientn(name = "FDR",
                          colors = c("blue","purple","red"),
                          space = "Lab", breaks = c(1e-25, 1e-17, 1e-10, 1e-6, 5e-2), labels = c("1e-25","1e-17","1e-10","1e-6","5e-2"),
                          limits = c(1e-30,5e-2), trans = "log") +
    scale_size_continuous("Gene Ratio",
                          breaks = seq(0.025, 0.6, 0.025),
                          labels = seq(0.025, 0.6, 0.025)) +
    ggtitle(plot_title) +
    theme_bw() +
    theme(text = element_text(size = 18),
          axis.text.x = element_text(angle = 0, hjust = 0.5),
          plot.title = element_text(hjust = 0.5, vjust = 2)))
  # if (class(p) == "try-error"){
  #   print(paste("no significant enrichment for"), name1, name2)
  # } else{
  #   ggsave(plot = p, filename, width = 9, height = 8)
  # }
  return(list(p, filename))
  
}



```




```{r}

compare_groups <- function(group1, group2, results_df, pval_cutoff = 0.34){
  name1 <- gsub("_and_", " ", group1) %>% gsub("-", " vs ", .) %>% sub(" none", "", .)
  name2 <- gsub("_and_", " ", group2) %>% gsub("-", " vs ", .) %>% sub(" none", "", .)
  name1 <- sub(" vs Parental", "", name1)
  name2 <- sub(" vs Parental", "", name2)
  N_features <- results_df %>% filter(contrast == group1) %>% nrow()
  all_features <- unique(results_df$feature)
  group1_df <- results_df %>%
    filter(contrast == group1,
           welch_adj < 0.05) %>%
    dplyr::select(feature, logFC)
  
  group2_df <- results_df %>%
    filter(contrast == group2,
           welch_adj < 0.05) %>%
    dplyr::select(feature, logFC)
  
  common_features <- intersect(group1_df$feature, group2_df$feature)
  pval <- hypergeometric_pval <- phyper(q = length(common_features), m = nrow(group1_df), 
                                        n = N_features - nrow(group1_df), k = nrow(group2_df), lower.tail = F)
  combined <- rbind(group1_df %>% filter(feature %in% common_features), 
                    group2_df %>% filter(feature %in% common_features)) %>%
    as.data.frame() %>%
    group_by(feature) %>%
    mutate(logFC = sign(logFC), 
           agree = sign(prod(logFC)) == 1)
  
  disagree <- combined %>% filter(!agree) %>%
    pull(feature) %>% unique()
  
  disagree_df <- rbind(group1_df %>% filter(feature %in% disagree) %>% mutate(group = group1),
                      group2_df %>% filter(feature %in% disagree) %>% mutate(group = group2)) %>%
    mutate(group = sub("-Parental_and_none$", "", group),
           group = gsub("_", " ", group),
           color = case_when(logFC > 0 ~ "red",
                             logFC < 0 ~ "blue")) 
  
  disagree_plot <- ggplot(disagree_df, aes(x = logFC, y = feature, fill = color)) + 
    geom_bar(stat = 'identity') + 
    facet_grid(. ~ group) + 
    guides(fill = "none") #+ theme(text = element_text(size = 21))
  
  ggsave(plot = disagree_plot, filename = paste0("disagree_plot_", group1, "_", group2, ".png"))
  
  combined <- combined %>%
    filter(agree) %>%
    dplyr::slice(1)


  venn.diagram(
    x = list(group1_df$feature, group2_df$feature),
    category.names = c(name1, name2),
    filename = paste("overlap_venn", gsub(" ", "_", tolower(name1)), "and",
                     gsub(" ", "_", tolower(name2)), "vs_parental_global.png", sep = "_"),
    output = FALSE,

    # Output features
    imagetype="png" ,
    height = 700 ,
    width = 700,
    resolution = 300,
    lwd = 1,
    col=c("#440154ff", '#21908dff'),
    fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3)),
    cex = 0.65,
    fontfamily = "sans",
    cat.cex = 0.7,
    cat.default.pos = "outer",
    cat.pos = c(-165, 15),
    cat.dist = c(0.05, 0.05),
    cat.col = c("#440154ff", '#21908dff'),
    main = paste0("pvalue = ", pval),
    main.cex = 0.69
  )
  plot_title <- paste("Genes active in", sub(" vs Parental", "", name1), "and", sub(" vs Parental", "", name2), "vs Parental")
  filename = paste0("over_representation_gobp_", gsub(" ", "_", tolower(plot_title)), ".png")

  clustList <- list('Over-expressed' = combined %>% filter(logFC == 1) %>% pull(feature),
                    'Under-expressed' = combined %>% filter(logFC == -1) %>% pull(feature),
                    'All features' = all_features)

  clustList <- lapply(clustList, intersect, mappedRkeys(org.Hs.egSYMBOL2EG))
  clustEG <- lapply(clustList, function(x) sapply(as.list(org.Hs.egSYMBOL2EG[x]), `[`, 1))

  xx <- compareCluster(clustEG[1:2], fun="enrichGO", OrgDb = org.Hs.eg.db, ont = "BP",
                       pvalueCutoff = pval_cutoff,
                       qvalueCutoff = pval_cutoff,
                       pAdjustMethod = "BH",
                       universe=clustEG[[3]], readable = T)
  # plot(xx, colorBy="qvalue")
  p <- try(dotplot(xx, showCategory = 7) +
    scale_color_gradientn(name = "FDR",
                          colors = c("blue","purple","red"),
                          space = "Lab", breaks = c(1e-25, 1e-17, 1e-10, 1e-6, 5e-2), labels = c("1e-25","1e-17","1e-10","1e-6","5e-2"),
                          limits = c(1e-30,5e-2), trans = "log") +
    scale_size_continuous("Gene Ratio",
                          breaks = seq(0.025, 0.6, 0.025),
                          labels = seq(0.025, 0.6, 0.025)) +
    ggtitle(plot_title) +
    theme_bw() +
    theme(text = element_text(size = 18),
          axis.text.x = element_text(angle = 0, hjust = 0.5),
          plot.title = element_text(hjust = 0.5, vjust = 2)))
  if (class(p) == "try-error"){
    print(paste("no significant enrichment for"), name1, name2)
  } else{
    try(ggsave(plot = p, filename, width = 9, height = 8))
  }
}


```


Global intersections

```{r}
group1 = "Early_and_G-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"

compare_groups(group1, group2, diffexp_results_global, pval_cutoff = 0.2)

group1 = "Early_and_GD-Parental_and_none"
group2 = "Late_and_GD-Parental_and_none"

## All signs agree
compare_groups(group1, group2, diffexp_results_global)


group1 = "Early_and_GV-Parental_and_none"
group2 = "Late_and_GV-Parental_and_none"

## All signs agree
compare_groups(group1, group2, diffexp_results_global)


group1 = "Early_and_GVD-Parental_and_none"
group2 = "Late_and_GVD-Parental_and_none"

## All signs agree
compare_groups(group1, group2, diffexp_results_global)


```


```{r}
group1 = "Early_and_GD-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

## Lowest pvalue for over-expressed is 0.51.
compare_groups(group1, group2, diffexp_results_global)



group1 = "Early_and_GV-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

## No significant enrichment, lowest pvalues are 0.12 and 0.33
compare_groups(group1, group2, diffexp_results_global)



group1 = "Early_and_GVD-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

## No significant enrichment, lowest pvalue is 0.4. No plot from here!
compare_groups(group1, group2, diffexp_results_global)


```


```{r}
group1 = "Late_and_GD-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"

compare_groups(group1, group2, diffexp_results_global)

group1 = "Late_and_GV-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"

compare_groups(group1, group2, diffexp_results_global)

group1 = "Late_and_GVD-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"

compare_groups(group1, group2, diffexp_results_global)


```


Phospho Intersections


```{r}
group1 = "Early_and_G-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

group1 = "Early_and_GD-Parental_and_none"
group2 = "Late_and_GD-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)


group1 = "Early_and_GV-Parental_and_none"
group2 = "Late_and_GV-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)


group1 = "Early_and_GVD-Parental_and_none"
group2 = "Late_and_GVD-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)


```



```{r}
group1 = "Early_and_GD-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

group1 = "Early_and_GV-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

group1 = "Early_and_GVD-Parental_and_none"
group2 = "Early_and_G-Parental_and_none"

p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)


```



```{r}
group1 = "Late_and_GD-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

group1 = "Late_and_GV-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"
## Only 6 phosphosites in common. No enrichment
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

group1 = "Late_and_GVD-Parental_and_none"
group2 = "Late_and_G-Parental_and_none"
p <- compare_groups_phospho(group1, group2, results_df = diffexp_results_phospho, pval_cutoff = 1)
ggsave(plot = p[[1]], filename = p[[2]], width = 9, height = 8)

```





